libs/core/algorithms/include/hpx/parallel/algorithms/nth_element.hpp

```diff
@@ -0,0 +1,411 @@
+            }
+            if (level == 0)
+            {
+                std::make_heap(first, end, comp);
+                std::sort_heap(first, nth, comp);
+                return;
+            };
+
+            // Filter the range and check which part contains the nth element
+            RandomIt c_last = filter(first, end, comp);
+
+            if (c_last == nth)
+                return;
+
+            if (nth < c_last)
+                nth_element_seq(first, nth, c_last, level - 1, comp, proj);
+            else
+                nth_element_seq(c_last + 1, nth, end, level - 1, comp, proj);
+
+            return;
+        };
+
+        template <typename Iter>
+        struct nth_element : public detail::algorithm<nth_element<Iter>, Iter>
+        {
+            nth_element()
+              : nth_element::algorithm("nth_element")
+            {
+            }
+
+            template <typename ExPolicy, typename RandomIt, typename Sent,
+                typename Pred, typename Proj>
+            static RandomIt sequential(ExPolicy, RandomIt first, RandomIt nth,
+                Sent last, Pred&& pred, Proj&& proj)
+            {
+                auto end = detail::advance_to_sentinel(first, last);
+
+                auto nelem = end - first;
+                if (nelem == 0)
+                    return first;
+
+                uint32_t level = detail::nbits64(nelem) * 2;
+                detail::nth_element_seq(first, nth, end, level,
+                    std::forward<Pred>(pred), std::forward<Proj>(proj));
+
+                return end;
+            }
+
+            template <typename ExPolicy, typename RandomIt, typename Sent,
+                typename Pred, typename Proj>
+            static typename util::detail::algorithm_result<ExPolicy,
+                RandomIt>::type
+            parallel(ExPolicy&& policy, RandomIt first, RandomIt nth, Sent last,
+                Pred&& pred, Proj&& proj)
+            {
+                using value_type =
+                    typename std::iterator_traits<RandomIt>::value_type;
+
+                RandomIt partitionIter, return_last;
+
+                if (first == last)
+                {
+                    return util::detail::algorithm_result<ExPolicy,
+                        RandomIt>::get(std::move(first));
+                }
+
+                if (nth == last)
+                {
+                    return util::detail::algorithm_result<ExPolicy,
+                        RandomIt>::get(std::move(nth));
+                }
+
+                try
+                {
+                    RandomIt last_iter =
+                        detail::advance_to_sentinel(first, last);
+                    return_last = last_iter;
+
+                    while (first != last_iter)
+                    {
+                        auto n = detail::distance(first, last_iter);
+
+                        // get random pivot index
+                        auto pivotIndex = std::rand() % n;
```

Comment:
Please note that `std::rand()` is not thread-safe. Could we use more sophisticated pivoting techniques instead (like in `hpx::stable_sort`)?

Reply:
Oh ok sure, thanks!